### DISCUSSION TITLE
Menu item positioning

### DISCUSSION THREAD

**Discussion by CyprinusCarpio** [fltk/fltk#1164] [Development]
One of the things I think could improve FLTK looks and readability is changing the positioning of menu items.

Currently, toggle and radio items widths are expanded by `FL_NORMAL_SIZE`. But, their labels appear to be offset to the right by a magic number of 3 (on top of `FL_NORMAL_SIZE`), without expanding the actual item width, which makes it so that the label is too close to the menu edge if that item is the widest:
![bez tytułu](https://github.com/user-attachments/assets/8029f738-abb8-4d6d-9bda-04921a3ea6fe)

To improve readability, I think standardizing the padding widths both left and right of the items would be best - i.e. remove the magic number label offset for radio and toggle items, offset normal menu items by `FL_NORMAL_SIZE` as well, and perhaps add `FL_NORMAL_SIZE` width padding to the right as well. It would look like so:
![bez tytułu](https://github.com/user-attachments/assets/b673c4e5-83bd-44eb-914a-765edeb040cd)

One downside of this approach is that some may take advantage of the fact that FLTK draws normal items without any (significant) padding on the left, for making "header" items and just using 4 or so spaces to add padding to "normal" items. Some solution for drawing "header" items like those would need to be found. And, all applications which add padding by using extra characters would need to be updated.

A less intrusive solution is to replace the magic number 3 by a width of some character, maybe a space or U+2005. This way a program dev can ensure consistent text offset for all menu items by prepending normal item labels with U+2001 and Space/U+2005 characters.

**Comment by MatthiasWM** [comment]
Thanks for the suggestions, @CyprinusCarpio . Which version of FLTK are you using?

Some of that can be implemented relatively easily, and not taking the check box offset into account when calculating maximum widget width is a bug that needs to be fixed. Apple uses the same indent for all items in a menu if at least one of the entries can have a check box.

Spacing should probably be based on the font height, or the maximum font height in a menu. 

The pulldown menu system is very high on the list for a complete rework in 1.5 .

**Reply by CyprinusCarpio** [reply]
I'm using 74d827f but the appearance is the same on all FLTK versions I used (starting with 1.3.8).
As I understand, the spacing is already based on font height (represented by `FL_NORMAL_SIZE` in `Fl_Menu_Item::measure` and `Fl_Menu_Item::draw`) but it's not consistent.

**Comment by MatthiasWM** [comment]
Thanks for the update. I asked because the 1.4 code is slightly different. I committed a fix to 1.4 master for the most pressing and actual bug of not measuring the gap between a possible checkbox and the label correctly. This does not fix your other layout observations.

As mentioned before, 1.4.1 is days from being released, so we only fix bugs now. The requested features will very likely be in 1.5 and possibly be backported to 1.4.x at some point.

**Comment by dannye** [comment]
FWIW, the fact that check/radio items and _non_-check/radio items are not aligned can be fixed by turning every non-check/radio item into a Fl_Multi_Label, where labela is a fully transparent image and labelb is the actual label:

![image](https://github.com/user-attachments/assets/0af8aebe-3aaf-4f9e-aa94-6c0911aab788)

But you can see in my case it's still off by 1 pixel. (I use a 16x16 transparent xpm but perhaps I should use a 17x17.)

Additionally, each of my labels end with a few trailing spaces, to ensure that the longest label ("Postprocess Channel" in this case) has a sufficient margin between it and its shortcut.

I think a 1.5 menu that had more alignment and more "elbow room" would be fantastic.

**Reply by MatthiasWM** [reply]
That look actually really nice! Thanks for "faking" the effect.

1.5 will have a completely new menu system (the old one will stick around) that uses regular widgets in groups instead of the list system. This will give complete control over the layout if wanted, but also allows for self aligning menu items. Dividers will have a true size, and menu items can have a different height if needed. And a lot more.
